### ISSUE TITLE
java.lang.ArrayIndexOutOfBoundsException

### ISSUE THREAD

**Issue by kzittritsch** [zendesk/maxwell#733]
Hi,
I am receiving the following error from Maxwell:
```
11:14:56,370 INFO  BinaryLogClient - Connected to 127.0.0.1:3306 at mysql-bin.000033/282296492 (sid:6379, cid:6834)
11:14:56,444 INFO  TaskManager - Stopping 3 tasks
11:14:56,445 ERROR TaskManager - cause:
java.lang.ArrayIndexOutOfBoundsException: 3
        at com.zendesk.maxwell.schema.columndef.EnumColumnDef.asString(EnumColumnDef.java:34) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.schema.columndef.EnumColumnDef.asJSON(EnumColumnDef.java:22) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.schema.columndef.EnumColumnDef.asJSON(EnumColumnDef.java:5) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.replication.BinlogConnectorEvent.writeData(BinlogConnectorEvent.java:88) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.replication.BinlogConnectorEvent.buildRowMap(BinlogConnectorEvent.java:136) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.replication.BinlogConnectorEvent.jsonMaps(BinlogConnectorEvent.java:163) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.replication.BinlogConnectorReplicator.getTransactionRows(BinlogConnectorReplicator.java:160) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.replication.BinlogConnectorReplicator.getRow(BinlogConnectorReplicator.java:271) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.replication.AbstractReplicator.work(AbstractReplicator.java:157) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:27) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:183) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.Maxwell.start(Maxwell.java:134) ~[maxwell-1.10.6.jar:1.10.6]
        at com.zendesk.maxwell.Maxwell.main(Maxwell.java:204) ~[maxwell-1.10.6.jar:1.10.6]
11:14:56,448 INFO  TaskManager - Stopping: com.zendesk.maxwell.producer.MaxwellKafkaProducerWorker@14c4965b
11:14:56,449 INFO  StoppableTaskState - MaxwellKafkaProducerWorker requestStop() called (in state: RUNNING)
11:14:56,449 INFO  KafkaProducer - Closing the Kafka producer with timeoutMillis = 9223372036854775807 ms.
11:14:56,450 DEBUG Sender - Beginning shutdown of Kafka producer I/O thread, sending remaining records.
11:14:56,451 DEBUG Metrics - Removed sensor with name connections-closed:client-id-producer-1
11:14:56,451 DEBUG Metrics - Removed sensor with name connections-created:client-id-producer-1
11:14:56,451 DEBUG Metrics - Removed sensor with name bytes-sent-received:client-id-producer-1
11:14:56,451 DEBUG Metrics - Removed sensor with name bytes-sent:client-id-producer-1
11:14:56,452 DEBUG Metrics - Removed sensor with name bytes-received:client-id-producer-1
11:14:56,452 DEBUG Metrics - Removed sensor with name select-time:client-id-producer-1
11:14:56,452 DEBUG Metrics - Removed sensor with name io-time:client-id-producer-1
11:14:56,452 DEBUG Sender - Shutdown of Kafka producer I/O thread has completed.
11:14:56,452 DEBUG KafkaProducer - The Kafka producer has closed.
11:14:56,453 INFO  TaskManager - Stopping: com.zendesk.maxwell.schema.PositionStoreThread@4355a2b9
11:14:56,453 INFO  StoppableTaskState - com.zendesk.maxwell.schema.PositionStoreThread requestStop() called (in state: RUNNING)
11:14:56,453 INFO  TaskManager - Stopping: com.zendesk.maxwell.replication.BinlogConnectorReplicator@11f38ba
11:14:56,453 INFO  StoppableTaskState - com.zendesk.maxwell.replication.BinlogConnectorReplicator requestStop() called (in state: STOPPED)
11:14:56,453 DEBUG TaskManager - Awaiting stop of: com.zendesk.maxwell.producer.MaxwellKafkaProducerWorker@14c4965b
11:14:56,454 DEBUG MysqlPositionStore - writing heartbeat: 1503054896453 (last heartbeat written: 1503054852217)
11:14:57,561 DEBUG TaskManager - Awaiting stop of: com.zendesk.maxwell.schema.PositionStoreThread@4355a2b9
11:14:57,662 DEBUG TaskManager - Awaiting stop of: com.zendesk.maxwell.replication.BinlogConnectorReplicator@11f38ba
java.lang.ArrayIndexOutOfBoundsException: 3
        at com.zendesk.maxwell.schema.columndef.EnumColumnDef.asString(EnumColumnDef.java:34)
        at com.zendesk.maxwell.schema.columndef.EnumColumnDef.asJSON(EnumColumnDef.java:22)
        at com.zendesk.maxwell.schema.columndef.EnumColumnDef.asJSON(EnumColumnDef.java:5)
        at com.zendesk.maxwell.replication.BinlogConnectorEvent.writeData(BinlogConnectorEvent.java:88)
        at com.zendesk.maxwell.replication.BinlogConnectorEvent.buildRowMap(BinlogConnectorEvent.java:136)
        at com.zendesk.maxwell.replication.BinlogConnectorEvent.jsonMaps(BinlogConnectorEvent.java:163)
        at com.zendesk.maxwell.replication.BinlogConnectorReplicator.getTransactionRows(BinlogConnectorReplicator.java:160)
        at com.zendesk.maxwell.replication.BinlogConnectorReplicator.getRow(BinlogConnectorReplicator.java:271)
        at com.zendesk.maxwell.replication.AbstractReplicator.work(AbstractReplicator.java:157)
        at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:27)
        at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:183)
        at com.zendesk.maxwell.Maxwell.start(Maxwell.java:134)
        at com.zendesk.maxwell.Maxwell.main(Maxwell.java:204)
```
I have dug into the binlog to view the statement at this position:
```
#170818  6:02:17 server id 10164428  end_log_pos 282296492 CRC32 0x82a6cb30     Write_rows: table id 459 flags: STMT_END_F
### INSERT INTO `database`.`table`
### SET
###   @1=11108661 /* INT meta=0 nullable=0 is_null=0 */
###   @2=94283 /* INT meta=0 nullable=0 is_null=0 */
###   @3=94865 /* INT meta=0 nullable=0 is_null=0 */
###   @4='0.00' /* VARSTRING(45) meta=45 nullable=0 is_null=0 */
###   @5='2017-08-18 06:02:17' /* DATETIME(0) meta=0 nullable=1 is_null=0 */
###   @6='787' /* VARSTRING(150) meta=150 nullable=1 is_null=0 */
###   @7='Enabled' /* VARSTRING(45) meta=45 nullable=1 is_null=0 */
# at 282296492
```
Can you please advise?

**Comment by kzittritsch** [comment]
FWIW we appear to have empty string `''` enum values in one of our tables. I resolved this by inserting "NULL" into the maxwell.columns table for those enum values.

**Comment by mortensen** [comment]
I had the same problem with our database. I could replicate and fix this issue with my commit (waiting for the pull to master). The bug is not in the EnumColumnDef.asString() method. It happens way before when the initial maxwell schema database is set up. It happens when you have enums with empty values like `my_enum` ENUM('','a','b','c'). The method extractEnumValues() in class SchemaCapturer uses the Apache StrTokenizer class. The parsed enum values would then be a, b and c (without the empty value). This later leads to the ArrayIndexOutOfBoundsException in EnumColumnDef.asString() due to different entry size and the "-1" done there.

**Comment by mortensen** [comment]
https://github.com/zendesk/maxwell/commit/73a8dfdf1e5338caa9ee56ca031044c5c56fedb8

**Comment by liulikun** [comment]
Fixed by #776
